src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ICollection.cs

```diff
@@ -53,5 +53,7 @@ bool IsReadOnly
         [DynamicDependency(nameof(Array.InternalArray__ICollection_Remove) + "``1", typeof(Array))]
 #endif
         bool Remove(T item);
+
+        int IReadOnlyCollection<T>.Count => Count;
```

Comment:
@eiriktsarpalis, if we end up keeping this inheritance, that will tip the scales for special-casing IReadOnlyXx in LINQ: in particular for places where we only need the source's count, we'll want to look for IReadOnlyCollection rather than ICollection, since every occurrence of the latter will also be the former.

Reply:
I believe it was decided to not make any changes in .NET 9 that would rely on this inheritance hierarchy change in case it needs to be reverted late in the cycle.

Reply:
Note that going to the default implemented method (that in turn calls another interface method) will involve a double interface dispatch. It remains to be seen how well (if at all) dynamic pgo will be able to get rid of it. The callsite in the default implementation is going to be megamorphic.

Reply:
Isn't this also something PGO can help with?

Reply:
The way PGO works is that it will collect all the possible types of `this` it saw at the callsite and then generate a cascade of `if (thisType.GetType() == typeof(Foo)) Foo.Method();`. There's a limit on how many checks can be generated that way, and each check has a cost. The more types you have that implement the ICollection but not IReadOnlyCollection, the worse it gets.

You'll typically not see this in microbenchmarks because those typically repeat the same thing over and over and dynamic PGO is really good at optimizing things that are repeated the same way over and over.

Reply:
The double dispatch is largely a point in time problem.

None of the built-in BCL types will actually hit it, because we provide an actual implementation. Downstream implementers are either already implementing both `ICollection<T>` and `IReadOnlyCollection<T>` or can be pointed towards doing so with an analyzer so they don't incur the double dispatch either.

Reply:
> I believe it was decided to not make any changes in .NET 9 that would rely on this inheritance hierarchy change in case it needs to be reverted late in the cycle.

But it could/would be done for .NET 10 (provided this change stays)?

Reply:
> > I believe it was decided to not make any changes in .NET 9 that would rely on this inheritance hierarchy change in case it needs to be reverted late in the cycle.
> 
> But it could/would be done for .NET 10 (provided this change stays)?

I don't agree with the stance and plan to submit a PR once this goes in. If we're confident enough to make this change at this point, we have to be fine with subsequent changes being based on it, as they absolutely will be whether intentionally or not. And it's no worse reverting one large PR and one small PR than one large PR of we decide we must.